### PR TITLE
Adding trigger to calculate some fields

### PR DESCRIPTION
**Pull request by drillinP**
* Ticket : #10 
*

### CODE REVIEW

File: pg_metadata/test/test_database_structure.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def test_install_database(self):
         }
         alg = "{}:create_database_structure".format(provider.id())
         results = processing.run(alg, params, feedback=self.feedback)
-        self.assertEqual('0.0.3', results['DATABASE_VERSION'])
+        self.assertEqual('0.0.4', results['DATABASE_VERSION'])
```
Comment (Gustry) [inline pg_metadata/test/test_database_structure.py:82]: Thanks for reminding this one. It has been fixed on master. You can rollback this change and rebase on master.

**Review by Gustry — COMMENTED** [review]
* Can you add a test ? There is a line layer in the database. Look in `test_sql.py` for an example.
* Does that work with a geometry less table ?

Can you rebase again on master ? I have included some tests on CI. Sorry for this.

**Comment by Gustry** [comment]
FYI, I have just added a geometry less table in the test database, if you rebase on master.

File: pg_metadata/install/sql/pgmetadata/10_FUNCTION.sql [hunk 2]
```diff
@@ -16,6 +16,35 @@ SET xmloption = content;
...
 SET row_security = off;
 
+-- calculate_fields_from_data()
+CREATE FUNCTION pgmetadata.calculate_fields_from_data() RETURNS trigger
+    LANGUAGE plpgsql
+    AS $$
+DECLARE
+    ma_table text;
+BEGIN
+	ma_table = CONCAT(NEW.schema_name, '.', NEW.table_name);
```
[the remark this answers is not in the document]
Comment (drillinP) [reply pg_metadata/install/sql/pgmetadata/10_FUNCTION.sql]: `table` it's a sql term. So this time we need prefix.

[the remark this answers is not in the document]
Comment (Gustry) [reply pg_metadata/install/sql/pgmetadata/10_FUNCTION.sql]: Ok !

File: pg_metadata/install/sql/pgmetadata/10_FUNCTION.sql [hunk 3]
```diff
@@ -16,6 +16,41 @@ SET xmloption = content;
...
+DECLARE
+    my_table text;
+    test_geom boolean;
+BEGIN
+	my_table = CONCAT(NEW.schema_name, '.', NEW.table_name);
+	EXECUTE 'SELECT COUNT(*) FROM ' || my_table
+	INTO NEW.feature_count;
+	EXECUTE 'SELECT true
+    FROM information_schema.columns 
+    WHERE table_schema=''' || NEW.schema_name || ''' AND table_name=''' || NEW.table_name || ''' and column_name=''geom'' ' INTO test_geom;
```
Comment (mdouchin) [inline pg_metadata/install/sql/pgmetadata/10_FUNCTION.sql:32]: To test if the table has a geometry column, I would use a direct SQL query on the `public.geometry_columns` table such as

```sql
SELECT * 
FROM geometry_columns WHERE f_table_schema = 'NOM_SCHEMA' AND f_table_name = 'NOM_TABLE' 
INTO geometry_columns_rec;
```
(this is an example, replace with the NEW.schema_name and table_name)
which will return nothing if no line is found, but a record with geometry_columns data if sucess.

If you modify the  `DECLARE` to add a new param: `geometry_columns_rec record;` which will receive the result of this query

This way you will be able to use directly the `geometry_columns_rec.geometry_type` like

```sql
NEW.geometry_type = geometry_columns_rec.geometry_type;
```

File: pg_metadata/install/sql/pgmetadata/20_TABLE_SEQUENCE_DEFAULT.sql [hunk 4]
```diff
@@ -71,7 +71,7 @@ CREATE TABLE pgmetadata.dataset (
     spatial_extent text,
     creation_date timestamp without time zone DEFAULT now() NOT NULL,
     update_date timestamp without time zone DEFAULT now(),
-    geom public.geometry(Polygon,4326)
+    geom public.geometry
```
Comment (mdouchin) [inline pg_metadata/install/sql/pgmetadata/20_TABLE_SEQUENCE_DEFAULT.sql:74]: I think we should keep use a geometry(Polygon,4326) here and use the `ST_Transform()` method to create the geometry from the table convex hull

File: pg_metadata/install/sql/upgrade/upgrade_to_0.0.4.sql [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+-- dataset trg_calculate_fields_from_data
+CREATE TRIGGER trg_calculate_fields_from_data
+    BEFORE INSERT OR UPDATE 
+    ON pgmetadata.dataset
+    FOR EACH ROW
+    EXECUTE PROCEDURE pgmetadata.calculate_fields_from_data();
+
+-- update geom field
+ALTER TABLE pgmetadata.dataset ALTER COLUMN geom TYPE geometry;
```
Comment (mdouchin) [inline pg_metadata/install/sql/upgrade/upgrade_to_0.0.4.sql:42]: Please keep geometry(Polygon, 4326) as explained above

**Review by mdouchin — CHANGES_REQUESTED** [review]

File: pg_metadata/test/test_sql.py [hunk 6]
```diff
@@ -36,3 +37,58 @@ def test_html_template(self):
...
+        # First with removing geom column
+        sql = "ALTER TABLE pgmetadata.lines DROP COLUMN geom"
+        self.connection.executeSql(sql)
+        sql = "UPDATE pgmetadata.dataset SET title = 'test after drop geom column'"
+        self.connection.executeSql(sql)
+        sql = "SELECT title, geometry_type, projection_authid, spatial_extent, geom FROM pgmetadata.dataset"
+        result = self.connection.executeSql(sql)
+        self.assertEqual(['test after drop geom column', NULL, NULL, NULL, NULL], result[0])
+
+        # Now with new table whithout geom column
```
Comment (Gustry) [inline pg_metadata/test/test_sql.py:79]:
```suggestion
    def test_trigger_calculate_fields_without_geom(self):
        """ Test the trigger on a geometry less table. """
```

@drillinP Il est conseillé de découper un test dés que le code qui suit n'a pas de lien avec le code précédent.
Comme ca, quand un test plante, cela ne stop pas les autres parties du code au sein du test (qui peut peut-être être OK)

File: pg_metadata/test/test_sql.py [hunk 7]
```diff
@@ -36,3 +37,58 @@ def test_html_template(self):
...
+        self.connection.executeSql(sql)
+        sql = "SELECT title, geometry_type, projection_authid, spatial_extent FROM pgmetadata.dataset"
+        result = self.connection.executeSql(sql)
+        self.assertEqual(
+            ['test lines title', 'LINESTRING', 'EPSG:4326', '3.854, 3.897, 43.5786, 43.622'],
+            result[0]
+        )
+
+        """ Test without geom column """
+        # First with removing geom column
```
Comment (Gustry) [inline pg_metadata/test/test_sql.py:70]:
```suggestion
        # Test with removing the geom column
```

@drillinP The is a difference between `""" Docstring """` and `# Comment`.

`""" Docstring... """` which must be unique (in a function) and at the top of the function. It's designed for documentation of the code and for display outside (when running tests for instance).

` # My comment` is designed for developpers, to give some comments about a specific block of code. It won't be displayed outside of the file or outside the code editor.

File: pg_metadata/test/test_sql.py [hunk 8]
```diff
@@ -36,3 +37,58 @@ def test_html_template(self):
                 "SELECT pgmetadata.get_dataset_item_html_content('pgmetadata','lines','main')")
         )
         self.assertEqual("<p>Test title</p><b>Test title</b>", result[0][0])
+
+    def test_trigger_calculate_fields(self):
+        """ Test if fields are correctly calculated """
+
+        """ Test with geom column """
```
Comment (Gustry) [inline pg_metadata/test/test_sql.py:44]:
```suggestion
    def test_trigger_calculate_fields_with_geom(self):
        """ Test if fields are correctly calculated on a layer having a geometry. """
```

**Review by Gustry — COMMENTED** [review]
I just reviewed the code about the Python side.

File: pg_metadata/test/test_sql.py [hunk 9]
```diff
@@ -36,3 +37,69 @@ def test_html_template(self):
...
+        # Test with removing the geom column
+        sql = "ALTER TABLE pgmetadata.lines DROP COLUMN geom"
+        self.connection.executeSql(sql)
+        sql = "UPDATE pgmetadata.dataset SET title = 'test after drop geom column'"
+        self.connection.executeSql(sql)
+        sql = "SELECT title, geometry_type, projection_authid, spatial_extent, geom FROM pgmetadata.dataset"
+        result = self.connection.executeSql(sql)
+        self.assertEqual(['test after drop geom column', NULL, NULL, NULL, NULL], result[0])
+
+        # Test with new table whithout geom column
```
Comment (Gustry) [inline pg_metadata/test/test_sql.py:90]: @drillinP C'est ici que je demandais de couper le code. Le code en-dessous est un nouveau test, car le code en-dessous ne dépend pas du code au dessus. En le découpant, on s'assure que le test est bien "unitaire" comme son nom l'indique.

```suggestion
        def test_trigger_calculate_fields_without_geom(self):
            """ Test compute fields on a geometry less table. """
```

File: pg_metadata/test/test_sql.py [hunk 10]
```diff
@@ -36,3 +37,69 @@ def test_html_template(self):
...
+        self.connection.executeSql(sql)
+        sql = "SELECT title, geometry_type, projection_authid, spatial_extent FROM pgmetadata.dataset"
+        result = self.connection.executeSql(sql)
+        self.assertEqual(
+            ['test lines title', 'LINESTRING', 'EPSG:4326', '3.854, 3.897, 43.5786, 43.622'],
+            result[0]
+        )
+
+    def test_trigger_calculate_fields_without_geom(self):
+        """ Test the trigger on a geometry less table. """
```
Comment (Gustry) [inline pg_metadata/test/test_sql.py:70]: Je pensais à cette coupure, mais je l'avais pas demandé, car c'était l'autre coupure. ;-)
Gardons la maintenant que c'est fait, mais par contre il faut renommer avec mon autre suggestion.

```suggestion
    def test_trigger_calculate_fields_when_removing_geom(self):
        """ Test the trigger on an existing table where we remove the geometry. """
```

**Review by Gustry — COMMENTED** [review]
LGTM for 🐍 (Python snake), I will let @mdouchin review the SQL 🐘
I read the SQL, it looks OK to me, but I haven't tried it.

**Comment by mdouchin** [comment]
@drillinP @Gustry I have 

* rewritten the SQL trigger function to always calculate the fields with RAISE NOTICE & better code readability
* added a function `pgmetadata.refresh_dataset_calculated_fields()` to manually refresh all dataset features
* I need to modify the upgrade SQL file !

**Comment by mdouchin** [comment]
Seems ok for me now :)

**Comment by mdouchin** [comment]
Manually rebasing -> close

**Pull request closed without merging**